src/cache-manager.js

```diff
@@ -50,22 +50,39 @@ class CacheManager {
     assert(rawUrl);
     this.debug(`putting ${rawUrl}`);
 
+    let mutex = new Mutex(this.redis, this.lockKey(rawUrl));
```

Comment:
So I'm no fan of two locks... anyways let's talk tomorrow :)

> My experience is that trusting the S3 docs is not always possible.

You have to read the reference docs carefully, but they are reliable. S3 gives some very specific consistency guarantees, understanding the specifics is hard. But those guarantees aren't soft.